### PR TITLE
Add tf dataset

### PR DESCRIPTION
**Pull request by awolant**
#### Why we need this PR?
- It adds new feature needed to improve DALI TensorFlow support and prepare it for the upcomming TF 2.0

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
To support future versions of TensorFlow there is a need to use tf.data.Dataset API rather then the older options of feeding the data to the TensorFlow graph.
This PR adds initial support for this API in DALI.

 - What was changed, added, removed?
Added `DALIDatasetOp` class and its Python wrappers. Modified how DALI TF plugin is build to include those.
 - What is most important part that reviewers should focus on?
How DALI API is used in the context of TensorFlow DatasetOp.

 - Was this PR tested? How?
Added python tests for `DALIDataset`.
 - Were docs and examples updated, if necessary?
Yes

**JIRA TASK**: [DALI-380]

### CODE REVIEW

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/tf.py [hunk 1]
```diff
@@ -115,6 +128,73 @@ def DALIIterator():
 def DALIRawIterator():
     return _dali_tf
 
+def get_tf_minor_version():
+  return tf.__version__.split('.')[1]
+
+
+class DALIDatasetV2(dataset_ops.DatasetSource):
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py]: how about having two separate modules (`v1.DALIDataset`, and `v2.DALIDataset`)

File: dali/python/nvidia/dali/plugin/tf.py [hunk 2]
```diff
@@ -115,6 +128,73 @@ def DALIIterator():
...
+    self._cpu_prefetch_queue_depth = cpu_prefetch_queue_depth
+    self._gpu_prefetch_queue_depth = gpu_prefetch_queue_depth
+    self._shapes = tuple(tf.TensorShape(shape) for shape in shapes)
+    self._dtypes = tuple(dtype for dtype in dtypes)
+
+    self._structure = structure.convert_legacy_structure(
+      self._dtypes, self._shapes, output_classes)
+
+    if get_tf_minor_version() == '14':
+      print('Using TF 14 veriosn')
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py]:
```suggestion
      print('Using TF 14 version')
```

File: dali/python/nvidia/dali/plugin/tf.py [hunk 3]
```diff
@@ -115,6 +128,73 @@ def DALIIterator():
...
+      super(DALIDatasetV2, self).__init__(self._as_variant_tensor())
+    else:
+      print('Using TF 13 veriosn')
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py]:
```suggestion
      print('Using TF 13 version')
```

File: dali_tf_plugin/dali_dataset_op.cc [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+      }
+    }
+    return Status::OK();
+  });
+
+}  // namespace
+}  // namespace data
+}  // namespace tensorflow
+
+#endif  // TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION >= 12
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_dataset_op.cc]:
```suggestion
#endif  // TF_MAJOR_VERSION == 1 && TF_MINOR_VERSION >= 13
```

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 5]
```diff
@@ -128,9 +128,9 @@ def install(self):
         dali_cflags, dali_lflags = get_dali_build_flags()
         tf_cflags, tf_lflags = get_tf_build_flags()
         cuda_cflags, cuda_lflags = get_cuda_build_flags()
-        plugin_src = self.src_path + '/daliop.cc'
+        plugin_src = self.src_path + '/daliop.cc' + ' ' + self.src_path + '/dali_dataset_op.cc'
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py]: maybe create a list of filenames and then construct this in a more generic way:
```for filename in filenames:
   plugin_src = plugin_src + ' ' + self.src_path + '/' + filename
```

File: dali/python/nvidia/dali/plugin/tf.py (hunk 1 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: This is sort of an idiom in the Tensorflow codebase (e.x. [`RandomDataset`](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/data/experimental/ops/random_ops.py#L32)) and it is there for compatibility reasons. When we will start to support TF 2.0 we will need it.
We could call them `*Wrapper` and `*Impl` for now but I wanted to do it like they do.

File: dali_tf_plugin/dali_dataset_op.cc (hunk 4 above)
Comment (awolant) [reply dali_tf_plugin/dali_dataset_op.cc]: Done

File: dali/python/nvidia/dali/plugin/tf.py (hunk 3 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: I removed those strings altogether to keep the output cleaner.

File: dali_tf_plugin/dali_tf_plugin_install_tool.py (hunk 5 above)
Comment (awolant) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py]: Done

File: dali/python/nvidia/dali/plugin/tf.py (hunk 2 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: I removed those strings altogether to keep the output cleaner.

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/tf.py [hunk 6]
```diff
@@ -116,5 +129,150 @@ def DALIRawIterator():
...
+      Number of CPU threads used by the pipeline.
+  `device_id` : int,
+      Id of GPU used by the pipeline.
+  `exec_separated` : bool,
+      Whether to execute the pipeline in a way that enables
+      overlapping CPU and GPU computation, typically resulting
+      in faster execution speed, but larger memory consumption.
+  `prefetch_queue_depth` : int,
+      depth of the executor queue. Deeper queue makes DALI more 
+      resistant to uneven execution  time of each batch, but it also
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py]:
```suggestion
      resistant to uneven execution time of each batch, but it also 
```

File: dali/python/nvidia/dali/plugin/tf.py [hunk 7]
```diff
@@ -116,5 +129,150 @@ def DALIRawIterator():
...
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py]:
```suggestion
      resistant to uneven execution time of each batch, but it also 
```

File: dali_tf_plugin/dali_dataset_op.cc [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    FUNC;                                                                  \
+  } catch (std::exception &e) {                                            \
+    std::string error = "DALI " + std::string(#FUNC)                       \
+                        + " failed: " + std::string(e.what());             \
+    std::cout << error << std::endl;                                       \
+    return ::tensorflow::errors::Internal(error);                          \
+  }                                                                        \
+} while (0)
+
+namespace tensorflow {
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_dataset_op.cc]: no `namespace dali` here?

File: dali_tf_plugin/dali_dataset_op.cc [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+      const int batch_size,
+      const int num_threads,
+      const int device_id,
+      const bool exec_separated,
+      const int prefetch_queue_depth,
+      const int cpu_prefetch_queue_depth,
+      const int gpu_prefetch_queue_depth,
+      const std::vector<PartialTensorShape> &shapes,
+      const DataTypeVector &dtypes) 
+      : DatasetBase(DatasetContext(context)),
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_dataset_op.cc]: nitpick: indent here

File: dali_tf_plugin/dali_dataset_op.cc [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    const std::vector<PartialTensorShape> &output_shapes() const override {
+      return shapes_;
+    }
+
+    string DebugString() const override { 
+      return "DALI::DatasetOp()::Dataset"; }
+
+    int64 Cardinality() const override { return kInfiniteCardinality; }
+
+    ~Dataset() {
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_dataset_op.cc]: virtual if you intend to use this as a base class for anything else

File: dali/python/nvidia/dali/plugin/tf.py (hunk 7 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: Done

File: dali_tf_plugin/dali_dataset_op.cc (hunk 8 above)
Comment (awolant) [reply dali_tf_plugin/dali_dataset_op.cc]: Done, removed `tensorflow` namespace.

File: dali_tf_plugin/dali_dataset_op.cc (hunk 9 above)
Comment (awolant) [reply dali_tf_plugin/dali_dataset_op.cc]: Done

File: dali/python/nvidia/dali/plugin/tf.py (hunk 6 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: Done

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/tf.py [hunk 11]
```diff
@@ -116,5 +117,136 @@ def DALIRawIterator():
...
+  return tf.__version__.split('.')[1]
+
+
+if _get_tf_minor_version() in {'13', '14'}:
+  from tensorflow.python.framework import ops
+  from tensorflow.python.data.ops import dataset_ops
+  from tensorflow.python.data.util import structure
+  import functools
+
+  class _DALIDatasetV2(dataset_ops.DatasetSource):
```
Comment (mzient) [inline dali/python/nvidia/dali/plugin/tf.py]: Out of sheer curiosity: why V2?

File: dali/python/nvidia/dali/plugin/tf.py [hunk 12]
```diff
@@ -116,5 +117,136 @@ def DALIRawIterator():
...
+
+      self._structure = structure.convert_legacy_structure(
+        self._dtypes, self._shapes, output_classes)
+
+      if _get_tf_minor_version() == '14':
+        super(_DALIDatasetV2, self).__init__(self._as_variant_tensor())
+      elif _get_tf_minor_version() == '13':
+        super(_DALIDatasetV2, self).__init__()
+      else:
+        raise RuntimeError('Unsupported TensorFlow version detected at runtime. DALIDataset supports versions: 1.13, 1.14')
```
Comment (JanuszL) [inline dali/python/nvidia/dali/plugin/tf.py]:
```suggestion
        raise RuntimeError('Unsupported TensorFlow version detected at runtime. DALIDataset supports versions: >= 1.13')
```
?

File: dali/test/python/test_dali_tf_dataset.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        standalone_results.append(
+            tuple(result.as_array() for result in standalone_pipeline.run()))
+        
+    for dataset_result, standalone_result in zip(dataset_results, standalone_results):
+        for dataset_out, standalone_out in zip(dataset_result, standalone_result):
+            assert np.array_equal(dataset_out, standalone_out)
+
+
+@raises(Exception)
+def test_differnt_num_shapes_dtypes():
```
Comment (JanuszL) [inline dali/test/python/test_dali_tf_dataset.py]: I think it is overlaps with the `test_tf_dataset`.

File: dali/test/python/test_dali_tf_dataset.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+    dataset_pipeline = TestPipeline(batch_size, num_threads)
+    shapes = [
+        (batch_size, 3, 224, 224), 
+        (batch_size, 1),
+        (batch_size, 1)]
+    dtypes = [
+        tf.float32,
+        tf.int32, 
+        tf.float32]
```
Comment (JanuszL) [inline dali/test/python/test_dali_tf_dataset.py]: Maybe it could be fp16 just to test it.

File: dali_tf_plugin/build_dali_tf.sh [hunk 15]
```diff
@@ -13,4 +13,4 @@ TF_CFLAGS=( $($PYTHON -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.g
 TF_LFLAGS=( $($PYTHON -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.get_link_flags()))') )
 
 
-$COMPILER -std=c++11 -O2 -shared -fPIC ${SRCS} -o ${LIB_NAME} ${INCL_DIRS} ${TF_CFLAGS[@]} ${TF_LFLAGS[@]} ${DALI_CFLAGS[@]} ${DALI_LFLAGS[@]}
+$COMPILER -std=c++11 -DNDEBUG -O2 -shared -fPIC ${SRCS} -o ${LIB_NAME} ${INCL_DIRS} ${TF_CFLAGS[@]} ${TF_LFLAGS[@]} ${DALI_CFLAGS[@]} ${DALI_LFLAGS[@]}
```
Comment (JanuszL) [inline dali_tf_plugin/build_dali_tf.sh]: Does `DNDEBUG` matter here?

Comment (JanuszL) [reply dali_tf_plugin/build_dali_tf.sh]: OK. I see https://github.com/tensorflow/tensorflow/issues/17316. Please add an appropriate comment here as well.

File: dali_tf_plugin/dali_dataset_op.cc [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+#include "tensorflow/core/framework/dataset.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/partial_tensor_shape.h"
+#include "tensorflow/core/framework/tensor.h"
+
+#define USE_TF_ALLOCATOR 0
+
+#if USE_TF_ALLOCATOR
+#include "tfallocator.h"
+#endif
```
Comment (JanuszL) [inline dali_tf_plugin/dali_dataset_op.cc]: I think we can remove it as it not used anyway.

Comment (awolant) [reply dali_tf_plugin/dali_dataset_op.cc]: Done

File: dali_tf_plugin/build_dali_tf.sh (hunk 15 above)
Comment (awolant) [reply dali_tf_plugin/build_dali_tf.sh]: Done

File: dali/test/python/test_dali_tf_dataset.py (hunk 14 above)
Comment (awolant) [reply dali/test/python/test_dali_tf_dataset.py]: Done

File: dali/test/python/test_dali_tf_dataset.py (hunk 13 above)
Comment (awolant) [reply dali/test/python/test_dali_tf_dataset.py]: The point of this was to see if assert will pick up the wrong arguments and throw an exception.

File: dali/python/nvidia/dali/plugin/tf.py (hunk 12 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: I only tested it for these two. >= would imply 1.15 and 2.0.

Maybe we can change it later, when we will test it properly?

File: dali/python/nvidia/dali/plugin/tf.py (hunk 11 above)
Comment (awolant) [reply dali/python/nvidia/dali/plugin/tf.py]: Normally in TensorFlow it looks like [this](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/data/experimental/ops/random_ops.py#L32). V2 will be exported to TensorFlow 2.0 and V1 to 1.x

For now I renamed V1 to `DALIDataset` to make the whole thing easier. When we will test it with TF 2.0 we will need V2.

**Comment by awolant** [comment]
!build

File: dali/python/nvidia/dali/plugin/tf.py (hunk 12 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/tf.py]: OK

**Pull request merged**